### PR TITLE
Faster and less memory hungry version of gdistances

### PR DESCRIPTION
**Pull request by bkamins**
#495 Here is the proposed code. I have moved the definitions to the end of the file as they use custom BFS implemenation.

Here is the performance testing code I used:
```
g = smallgraph(:tutte)
println("tutte 3")
@time a = gdistances(g,3);

srand(1)
for i in 0:5
    g = watts_strogatz(20000, 100, 0.5)
    vl = sample(1:100, 5i, replace=false)
    println("Watts-Strogatz ", vl)
    @time a = gdistances(g,vl);
end
```

It seems to use less time and memory than the current implementation. I have also updated tests for a corner case of emply `source` collection.

### CODE REVIEW

**Comment by sbromberger** [comment]
@bkamins - thanks. Could you post benchmarks for this in this thread? Specifically, try both sparse and dense graphs, with single- and multi-source. That way we'll have it in the record here and it'll be another verification.

@jpfairbanks take a look at this - even if we revert the regression in https://github.com/JuliaGraphs/LightGraphs.jl/commit/c571fdfdf9bee2a01c15dc989cdebfe750a9211f, @bkamins' code is 5x faster than using the generic bfs traversal.

**Comment by bkamins** [comment]
Here are the benchmark results:
```
tutte graph 50% source
old:  0.007651 seconds (1.89 k allocations: 81.509 KB)
new:  0.000004 seconds (2 allocations: 896 bytes)

tutte graph single source
old:  0.000067 seconds (145 allocations: 5.516 KB)
new:  0.000003 seconds (2 allocations: 896 bytes)

large sparse graph 50% source
old:  4.211344 seconds (40.98 M allocations: 794.589 MB, 3.60% gc time)
new:  0.399337 seconds (4 allocations: 15.259 MB, 0.77% gc time)

large sparse graph single source
old:  4.230289 seconds (40.98 M allocations: 790.589 MB, 3.59% gc time)
new:  0.386659 seconds (4 allocations: 15.259 MB, 0.65% gc time)

medium dense graph 50% source
old: 30.604378 seconds (447.57 M allocations: 8.162 GB, 5.25% gc time)
new:  0.165119 seconds (4 allocations: 1.526 MB)

medium dense graph single source
old: 30.385432 seconds (447.57 M allocations: 8.160 GB, 5.27% gc time)
new:  0.140546 seconds (4 allocations: 1.526 MB)

small very dense graph 50% source
old:  3.042858 seconds (44.49 M allocations: 831.601 MB, 5.47% gc time)
new:  0.015054 seconds (4 allocations: 78.281 KB)

small very dense graph single source
old:  3.039986 seconds (44.49 M allocations: 831.601 MB, 5.47% gc time)
new:  0.014593 seconds (4 allocations: 78.281 KB)
```

and the code (old is current master, new is the proposed implementation):
```
using LightGraphs
using Distributions

function new_gdistances!(g::SimpleGraph, source, dists)
    n = nv(g)
    fill!(dists, -1)
    queue = fill(-1, n)
    for i in 1:length(source)
        queue[i] = source[i]
        dists[source[i]] = 0
    end
    head = 1
    tail = length(source)
    while head <= tail
        current = queue[head]
        distance = dists[current] + 1
        head += 1
        for j in fadj(g, current)
            if dists[j] == -1
                dists[j] = distance
                tail += 1
                queue[tail] = j
            end
        end
    end
    dists
end

function new_gdistances(g::SimpleGraph, source)
    new_gdistances!(g, source, Vector{Int}(nv(g)))
end

function run_test(g, multi, caption)
    vl = sample(1:nv(g),
                multi ? div(nv(g), 2) : 1,
                replace = false)
    println("\n", caption, multi ? " 50% source":" single source")
    print("old:")
    @time a = gdistances(g,vl);
    print("new:")
    @time b = new_gdistances(g,vl);
    @assert a == b
end

srand(1)
gs = [(smallgraph(:tutte), "tutte graph"),
      (watts_strogatz(10^6, 10, 1), "large sparse graph"),
      (watts_strogatz(10^5, 1000, 0), "medium dense graph"),
      (watts_strogatz(5000, 2000, 0.5), "small very dense graph")]
@assert gdistances(gs[1][1], 3) == new_gdistances(gs[1][1], 3)

for (g, c) in gs, m in [true, false]
    run_test(g, m, c)
end
```

**Comment by sbromberger** [comment]
Beautiful. Thank you for doing that extra work. Those numbers are compelling!

**Comment by jpfairbanks** [comment]
Can we add these benchmark cases to the tests? Up to 10^4 vertices for speed. Including the old definition in the test source code so that we can check for correctness? And add the full benchmarks to a file `/bench/gdistances.jl`

File: src/traversals/bfs.jl [hunk 1]
```diff
@@ -265,3 +234,49 @@ function bipartite_map(g::SimpleGraph)
...
+    n = nv(g)
+    fill!(dists, -1)
+    queue = Vector{Int}(n)
+    for i in 1:length(source)
+        queue[i] = source[i]
+        dists[source[i]] = 0
+    end
+    head = 1
+    tail = length(source)
+    while head <= tail
```
Comment (jpfairbanks) [inline src/traversals/bfs.jl]: This is a bespoke implementation of BFS can it be handled within the Visitor Framework?

Comment (jpfairbanks) [inline src/traversals/bfs.jl]: For example this BFS for connected components which is memory efficient. https://github.com/JuliaGraphs/LightGraphs.jl/blob/master/src/traversals/bfs.jl#L184

**Comment by sbromberger** [comment]
> Can we add these benchmark cases to the tests? Up to 10^4 vertices for speed. Including the old definition in the test source code so that we can check for correctness? And add the full benchmarks to a file /bench/gdistances.jl

Wait - no. I don't like this idea. We should do this when we get #45 implemented but there's no point in doing that now, especially adding the old code back in.

The old code was slower - originally by a factor of 5, but then we had a very bad regression in https://github.com/JuliaGraphs/LightGraphs.jl/commit/c571fdfdf9bee2a01c15dc989cdebfe750a9211f that caused a massive performance hit. This code is the fastest implementation yet.

File: src/traversals/bfs.jl (hunk 1 above)
Comment (sbromberger) [inline src/traversals/bfs.jl]: See https://github.com/JuliaGraphs/LightGraphs.jl/pull/496#issuecomment-257320324 and https://github.com/JuliaGraphs/LightGraphs.jl/issues/495#issuecomment-257190930 (et seq).

**Comment by jpfairbanks** [comment]
The code in master doesn't look like it would be that much slower than what is written here. Can we do some probing into why the code in master is so much slower? We could probably fix a performance bug that makes all the traversals faster. I'd rather fix a bug in the existing BFS traversal code, than use a custom BFS implementation.

**Comment by sbromberger** [comment]
The code in master overallocates since https://github.com/JuliaGraphs/LightGraphs.jl/commit/c571fdfdf9bee2a01c15dc989cdebfe750a9211f effectively got rid of `gdistances!`.

**Comment by jpfairbanks** [comment]
hmm... GdistanceVisitor never gets a definition for `examine_neighbor!`

**Comment by jpfairbanks** [comment]
> The  code in master overallocates since c571fdf effectively got rid of gdistances!.

Which array is overallocated?

**Comment by jpfairbanks** [comment]
Here is a basic profiling result based on @bkamins' benchmarkk
```julia
julia> g = watts_strogatz(10000, 2000, 0.5)
{10000, 10000000} undirected graph

julia> @profile gdistances(g, 3); Base.Profile.print()
4533 ./event.jl:68; (::Base.REPL.##3#4{Base.REPL.REPLBackend})()
 4533 ./REPL.jl:95; macro expansion
  4533 ./REPL.jl:64; eval_user_input(::Any, ::Base.REPL.REPLBackend)
   4533 ./boot.jl:234; eval(::Module, ::Any)
    4533 ./<missing>:?; anonymous
     4533 ./profile.jl:16; macro expansion;
      4533 ...ia/v0.5/LightGraphs/src/traversals/bfs.jl:113; gdistances(::LightGraphs.Graph, ::Int64)
       4533 ...ia/v0.5/LightGraphs/src/traversals/bfs.jl:99; gdistances!(::LightGraphs.Graph, ::Int64, ::Array{Int64,1})
        4533 ./<missing>:0; (::LightGraphs.#kw##traverse_graph!)(::Array{Any,1}, ::LightGr...
         4533 ...a/v0.5/LightGraphs/src/traversals/bfs.jl:73; #traverse_graph!#15(::Array{Int64,1}, ::LightGraphs.DummyEdg...
          2048 .../v0.5/LightGraphs/src/traversals/bfs.jl:40; breadth_first_visit_impl!(::LightGraphs.Graph, ::Array{Int64...
           14 ./array.jl:0; done(::Array{Int64,1}, ::Int64)
           4  ./array.jl:381; done(::Array{Int64,1}, ::Int64)
           56 ./array.jl:380; next(::Array{Int64,1}, ::Int64)
           7  ./bool.jl:0; !(::Bool)
           9  ./bool.jl:16; !(::Bool)
           1  ...mes/.julia/v0.5/LightGraphs/src/core.jl:243; out_neighbors(::LightGraphs.Graph, ::Int64)
          404  .../v0.5/LightGraphs/src/traversals/bfs.jl:41; breadth_first_visit_impl!(::LightGraphs.Graph, ::Array{Int64...
           4  ./abstractarray.jl:0; get(::Array{Int64,1}, ::Int64, ::Int64)
           26 ./abstractarray.jl:912; get(::Array{Int64,1}, ::Int64, ::Int64)
          1169 .../v0.5/LightGraphs/src/traversals/bfs.jl:42; breadth_first_visit_impl!(::LightGraphs.Graph, ::Array{Int64...
           8 ./essentials.jl:0; convert(::Type{Int64}, ::Int64)
          547  .../v0.5/LightGraphs/src/traversals/bfs.jl:44; breadth_first_visit_impl!(::LightGraphs.Graph, ::Array{Int64...
          322  .../v0.5/LightGraphs/src/traversals/bfs.jl:46; breadth_first_visit_impl!(::LightGraphs.Graph, ::Array{Int64...
           1  ./promotion.jl:0; ==(::Int64, ::Int64)
           10 ./promotion.jl:269; ==(::Int64, ::Int64)

```


This appears to be the offender:  `1169 .../v0.5/LightGraphs/src/traversals/bfs.jl:42; breadth_first_visit_impl!(::LightGraphs.Graph, ::Array{Int64...`

And Line 42 in the source is `v_edge = Edge(u,v)` 

Which is inconsistent with my conception of the Pair{Int,Int} type being put on the stack instead of the heap.

**Comment by jpfairbanks** [comment]
The only significant difference between the two codes, should be the `push!` on line 49 of bfs.jl. The implementation in this PR uses a fixed size queue. Which we could do in the Visitor methods currently implemented.

**Comment by sbromberger** [comment]
There are lots of things I don't like about our current visitor-based traversal. It's too complicated for me to figure out easily, for starters. But if you [look at `gdistances!`](https://github.com/JuliaGraphs/LightGraphs.jl/blob/master/src/traversals/bfs.jl#L97), as an example, we really need typing on the arguments.

In any case, if you can put together a version that matches performance, we'll do that instead.

**Comment by sbromberger** [comment]
@jpfairbanks - I'm inclined to merge this soon unless you think you can come up with a more general solution to the visitor inefficiency. Could you let me know?

**Comment by jpfairbanks** [comment]
@sbromberger can you comment more on what you don't like about the visitor traversal system we have? I am uncomfortable removing code that is slow without understanding why it is slow.

**Comment by sbromberger** [comment]
It's complicated and inconsistent, and (as we've discovered) it's slow. If we (you) can identify the source of slowness and fix this problem generally, then I'm ok with keeping traversals, but our priorities are correctness and speed, so if we have a faster algorithm, then we should be using it.

**Comment by bkamins** [comment]
@sbromberger @jpfairbanks Let me just add a perspective of LightGraphs user (and thank you both for great work on this package).

Current implementation in master does a lot of bookkeeping that is not required in `gdistances` case.
Eg. if you remove `v_edge` variable alone in `breadth_first_visit_impl!` function which is not needed for `gdistances` the speed goes up by ~30% and memory usage goes down by ~30%. I could imagine that in general in some cases it might be useful to track the edges, but e.g. currently it is not used at all as `examine_neighbor!` does not use `ecolor` argument anyway if I understand the code correctly.

What I would see as the best option for a LightGraphs user:
1) maximally efficient `gdistances` as this is the function that will be used 80% of time in practice;
2) a genral BFS traversal routine for 20% of special cases.

As a user I do not care if `gdistances` uses generic BFS or not (but probably you, as package developers, do care).

**Comment by sbromberger** [comment]
> What I would see as the best option for a LightGraphs user:
1) maximally efficient gdistances as this is the function that will be used 80% of time in practice;
2) a genral BFS traversal routine for 20% of special cases.

This is precisely what I'd like to see as well. That is: generalized algorithms that fare no worse than any other – until, like in this case, that changes - at which time we prefer a specialized algorithm that's API- and feature-compatible with the generalized one.

**Comment by sbromberger** [comment]
and @bkamins - I'm sorry this is taking so long, but the discussion is very important at this point. Your code has brought a few issues to the forefront that need some resolution. Stick with us, please.

**Comment by jpfairbanks** [comment]
Ok then lets merge this improvement and open an issue for improving the performance of the general code.

**Comment by jpfairbanks** [comment]
issue: #497

**Comment by sbromberger** [comment]
@bkamins thank you for this code. If you have time, please help us out with #497 :)
